### PR TITLE
Multiple roles

### PR DESCRIPTION
**Pull request by riley**
## What does this PR do?

This PR started out as adding a `STAFF` role, but ended up being a re-arch of how permissions are handled in Talk.

There is now a STAFF level of user which cannot do most things. Also, only Admins can
- view email addresses
- update config settings

## How do I test this PR?

- have 4 different users
  - `ADMIN`, `MODERATOR`, `STAFF` and `""`
- they should be able to (or not able to) do all the things outlined in [the feature](https://www.pivotaltracker.com/story/show/144717969)

## still TODO
- show/hide emails on the Community page for moderators
- show/hide roles dropdowns for non-admins
this will be broken out into a separate PR

### CODE REVIEW

File: models/user.js [hunk 1]
```diff
@@ -157,14 +159,74 @@ UserSchema.index({
...
+UserSchema.method('canChangeUserRoles', function () {
+  return !!intersection(['ADMIN', 'MODERATOR'], this.roles).length;
+});
+
+UserSchema.method('canSetCommentStatus', function () {
+  return !!intersection(['ADMIN', 'MODERATOR'], this.roles).length;
+});
+
+UserSchema.method('canSetUserStatus', function () {
+  return !!intersection(['ADMIN', 'MODERATOR'], this.roles).length;
```
Comment (wyattjoh) [inline models/user.js]: Just leaving a note about pulling these functions out before it's merged :)

**Review by wyattjoh — CHANGES_REQUESTED** [review]

File: models/user.js [hunk 2]
```diff
@@ -157,14 +160,10 @@ UserSchema.index({
...
+ * returns true if a commenter is staff
  */
-UserSchema.method('hasRoles', function(...roles) {
-  return roles.every((role) => {
-
-    // TODO: remove toUpperCase() once we've migrated usage.
-    return this.roles.indexOf(role.toUpperCase()) >= 0;
-  });
+UserSchema.method('isStaff', function () {
+  return !!intersection(['ADMIN', 'MODERATOR', 'STAFF'], this.roles).length;
```
Comment (wyattjoh) [inline models/user.js]: Replace logical comparison with:

```diff
- return !!intersection(['ADMIN', 'MODERATOR', 'STAFF'], this.roles).length;
+ return intersection(['ADMIN', 'MODERATOR', 'STAFF'], this.roles).length !== 0;
```

File: perms/mutationReducer.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
+const intersection = require('lodash/intersection');
+const check = (user, roles) => {
```
Comment (wyattjoh) [inline perms/mutationReducer.js]: Abstract the `check` function into a util.js file in perms that is imported.

File: perms/queryReducer.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
+const intersection = require('lodash/intersection');
+const check = (user, roles) => {
```
Comment (wyattjoh) [inline perms/queryReducer.js]: Abstract the `check` function into a util.js file in perms that is imported.

File: perms/rootReducer.js [hunk 5]
```diff
@@ -0,0 +1,15 @@
+module.exports = {
+  DUMMY_ROLE: 'DUMMY_ROLE',
```
Comment (wyattjoh) [inline perms/rootReducer.js]: Remove `DUMMY_ROLE`

File: perms/rootReducer.js [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+  checkRoles: function (user, perm) {
+
+    // this runs before everything
+    if (user.status === 'BANNED') {
+      return false;
+    }
+
+    switch (perm) {
```
Comment (wyattjoh) [inline perms/rootReducer.js]: Remove empty switch statement.

File: services/users.js [hunk 7]
```diff
@@ -389,13 +389,10 @@ module.exports = class UsersService {
...
-      id: id
-    }, {
-      $addToSet: {
-        roles: role
-      }
-    });
+    // 5.11.2017 - Restricting this to a hierarchical system like WordPress
+    // where you can only set one role at a time.
+    // I'm not changing the data structure here, because I don't want a migration
+    return UserModel.update({id}, {$set: {roles: [role]}});
```
Comment (wyattjoh) [inline services/users.js]: Remove comment.

**Review by wyattjoh — CHANGES_REQUESTED** [review]

**Review by wyattjoh — APPROVED** [review]

**Review by alinhle — CHANGES_REQUESTED** [review]
Don't get me wrong, this all looks great 👍  Just want to sync with Riley about handling suspended and banned users before we merge this.

**Review by alinhle — APPROVED** [review]

**Pull request merged**